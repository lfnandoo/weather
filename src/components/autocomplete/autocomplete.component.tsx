import { ChangeEvent, InputHTMLAttributes, useEffect, useRef, useState } from 'react';
import { useDebounce } from '../../hooks/use-debounce.hook';
import { AutocompleteInput } from './autocomplete.styles';
import { List } from './components/list/list.component';

interface Option {
  description: string;
  value: string;
  data: unknown;
}

interface AutocompleteProps extends Omit<InputHTMLAttributes<any>, 'onChange'> {
  onSearch: (searchTerm: string) => Promise<Option[]>;
  onChange: (value: Option) => void;
  debounce?: number;
}

function Autocomplete({ onSearch, onChange, debounce = 500, ...inputProps }: AutocompleteProps) {
  const searchRef = useRef({ enableSearch: false });
  const [isListOpen, setIsListOpen] = useState(false);
  const [isSearching, setIsSearching] = useState(false);
  const [searchTerm, setSearchTerm] = useState('');
  const [items, setItems] = useState([] as Option[]);
  const debouncedSearchTerm = useDebounce(searchTerm, debounce);

  async function handleSearch() {
    setIsSearching(true);
    setIsListOpen(true);
    const newItems = await onSearch(searchTerm);
    setItems(newItems);
    setIsSearching(false);
  }

  function handleChange(index: number) {
    setIsListOpen(false);
    const item = items[index];
    setSearchTerm(item.description);
    onChange(item);
    searchRef.current.enableSearch = false;
  }

  function handleInputChange(e: ChangeEvent<HTMLInputElement>) {
    setSearchTerm(e.target.value);
    searchRef.current.enableSearch = true;
  }

  useEffect(() => {
    if (debouncedSearchTerm && searchRef.current.enableSearch) {
      handleSearch();
    }
  }, [debouncedSearchTerm]);

  return (
    <>
      <AutocompleteInput
        {...inputProps}
        data-testid="autocomplete-input"
        type="text"
        value={searchTerm}
        onChange={handleInputChange}
        autoComplete="off"
      />
      {isListOpen && <List items={items} onClick={handleChange} isLoading={isSearching} />}
    </>
  );
}

export { Autocomplete };
