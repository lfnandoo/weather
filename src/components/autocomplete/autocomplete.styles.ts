import styled from 'styled-components';

const AutocompleteInput = styled.input`
  font-size: 20px;
  background-color: transparent;
  outline: none;
  border: 1px solid rgba(255, 255, 255, 0.2);
  border-radius: 20px;
  padding: 5px;
  width: calc(100% - 10px);
  font-weight: 100;
  color: rgba(0, 0, 0, 0.6);
`;

export { AutocompleteInput };
