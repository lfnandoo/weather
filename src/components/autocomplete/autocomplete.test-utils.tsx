import { fireEvent } from '@testing-library/dom';
import { act, render } from '@testing-library/react';
import { Autocomplete } from './autocomplete.component';

function autocompleteCreateAndChangeInput() {
  const onSearch = jest.fn(() => {
    return Promise.resolve([
      { description: 'S達o Paulo', value: '1', data: {} },
      { description: 'S達o Cristov達o', value: '2', data: {} },
    ]);
  });
  jest.useFakeTimers();
  const onChange = jest.fn();
  const renderResult = render(<Autocomplete onChange={onChange} onSearch={onSearch} />);

  fireEvent.change(renderResult.getByTestId('autocomplete-input'), { target: { value: 'S達o' } });

  act(() => {
    jest.advanceTimersByTime(500);
  });

  return { renderResult, onSearch, onChange };
}

export { autocompleteCreateAndChangeInput };
