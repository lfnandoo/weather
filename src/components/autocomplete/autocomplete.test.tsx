import { fireEvent, waitFor } from '@testing-library/dom';
import { autocompleteCreateAndChangeInput } from './autocomplete.test-utils';

describe('Autocomplete Component', () => {
  it('should trigger onSearch on input changes', async () => {
    const { onSearch } = autocompleteCreateAndChangeInput();

    await waitFor(() => {
      expect(onSearch).toHaveBeenCalledWith('S達o');
    });
  });

  it('should render suggestions list', async () => {
    const {
      onSearch,
      renderResult: { queryByTestId, queryAllByTestId },
    } = autocompleteCreateAndChangeInput();

    await waitFor(() => {
      expect(onSearch).toHaveBeenCalledWith('S達o');

      expect(queryByTestId('suggestions')).toBeInTheDocument();

      expect(queryAllByTestId('item')).toHaveLength(2);
    });
  });

  it('should trigger onChange when click on a suggestion', async () => {
    const {
      onSearch,
      onChange,
      renderResult: { queryByTestId, queryAllByTestId },
    } = autocompleteCreateAndChangeInput();

    await waitFor(() => {
      expect(onSearch).toHaveBeenCalledWith('S達o');

      expect(queryByTestId('suggestions')).toBeInTheDocument();

      const itemElements = queryAllByTestId('item');
      expect(itemElements).toHaveLength(2);

      fireEvent.click(itemElements[0]);
    });

    expect(onChange).toBeCalledWith({ description: 'S達o Paulo', value: '1', data: {} });
  });
});
