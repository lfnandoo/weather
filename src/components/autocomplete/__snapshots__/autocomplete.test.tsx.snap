// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Autocomplete Component should render correctly 1`] = `
<DocumentFragment>
  <input
    autocomplete="off"
    class="sc-bczRLJ lbLDEh"
    data-testid="autocomplete-input"
    type="text"
    value="SÃ£o"
  />
  <div
    class="sc-dkzDqf eevhgj"
  >
    <ul
      class="sc-hKMtZM iTymCl"
      data-testid="suggestions"
    >
      <li
        data-testid="item-loading"
      >
        <div
          class="sc-gsnTZi fIklhb"
          data-testid="skeleton"
          height="20px"
          width="150px"
        />
      </li>
      <li
        data-testid="item-loading"
      >
        <div
          class="sc-gsnTZi fIklhb"
          data-testid="skeleton"
          height="20px"
          width="150px"
        />
      </li>
      <li
        data-testid="item-loading"
      >
        <div
          class="sc-gsnTZi fIklhb"
          data-testid="skeleton"
          height="20px"
          width="150px"
        />
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
